### PR TITLE
Change dead struct indirection to nullcheck

### LINKED ISSUE #36237: Runtime crash during JIT register allocation in .NET 5
Running the following program on .NET 5 (and master) with optimizations enabled crashes the runtime during register allocation:
```csharp
struct S0
{
    public bool F7;
}

class C1
{
    public S0 F7;
}

public class Program
{
    static C1 s_8;
    public static void Main()
    {
        uint[] vr0 = default(uint[]);
        if (s_8.F7.F7)
        {
            vr0 = vr0;
        }
    }
}
```

In checked builds the following assertion fires:
```
Assert failure(PID 18840 [0x00004998], Thread: 18212 [0x4724]): Assertion failed 'indirTree->TypeGet() != TYP_STRUCT' in 'Program:Main()' during 'Linear scan register alloc' (IL size 22)

    File: D:\dev\dotnet\runtime\src\coreclr\src\jit\lsraxarch.cpp Line: 2763
    Image: D:\dev\dotnet\runtime\artifacts\tests\coreclr\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe
```

cc @CarolEidt

### PR DESCRIPTION
**Pull request by CarolEidt**
Fix #36237

### CODE REVIEW

**Comment by CarolEidt** [comment]
The test case was creating a dead `GT_FIELD` of `TYP_STRUCT` in `fgOptimizeBranchToNext()`. The code in the importer for `CEE_POP` was handling the case of a dead struct value, but other places were not. I extracted the code from there, and add a call to it from `gtExtractSideEffList()`, which required adding a `BasicBlock` operand to that method.
Most of the other callees of `gtExtractSideEffList()` will not currently be dealing with struct values, but they could be in future.

File: src/coreclr/src/jit/gentree.cpp [hunk 1]
```diff
@@ -15563,6 +15563,45 @@ GenTree* Compiler::gtBuildCommaList(GenTree* list, GenTree* expr)
     }
 }
 
+//------------------------------------------------------------------------
+// gtExtractSideEffList: Fixup a side-effect tree that produces a (dead) struct result
+//
+// Arguments:
+//    expr       - the side effect expression tree
```
Comment (AndyAyersMS) [inline src/coreclr/src/jit/gentree.cpp:15570]: Add comments for `block` and `clsHnd`?

Was puzzled why block was needed, but see it's apparently just to mark the block as containing a null check, which seems fine.

File: src/coreclr/tests/src/JIT/Regression/JitBlue/Runtime_36237/Runtime_36237.csproj [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+  </PropertyGroup>
+  <PropertyGroup>
+    <DebugType />
+    <Optimize>True</Optimize>
+    <AllowUnsafeBlocks>True</AllowUnsafeBlocks>
```
Comment (AndyAyersMS) [inline src/coreclr/tests/src/JIT/Regression/JitBlue/Runtime_36237/Runtime_36237.csproj:8]: Not needed (nothing unsafe in the test)?

**Review by AndyAyersMS — COMMENTED** [review]
Curious to learn what exactly goes wrong if we don't do this, but the fixes make sense.

Nice to see us consolidate around GT_NULLCHECK and then have that get optimized/optimally expanded later.

**Comment by CarolEidt** [comment]
@dotnet/dnceng - is there any way to re-run a cancelled job?

**Comment by MattGal** [comment]
> @dotnet/dnceng - is there any way to re-run a cancelled job?

Hi Carol.  Yes, if you click through all the way to Azure Devops, "Rerun Failed Jobs" in the top right will do this for you.  I've even seen it occasionally be available from the GitHub intermediate UI. 

I've already clicked this, sadly before screen shotting the button (I had to make sure it worked...) and the jobs are running now.  

Timeouts were from the SDK install script filling the temp dir with 20+ GB of dotnet SDKs on all the Ubuntu.1804.Amd64.Open machines, this should be fixed in the near future, but also the ASP.NET team has added cleanup of the random zip file name as part of their helix work items.

**Comment by CarolEidt** [comment]
Thanks @MattGal !

**Comment by CarolEidt** [comment]
Closing this in favor of a different fix.

**Pull request closed without merging**